File: dsa in python/merge_sort.py
result=[4,3,2,1]
def merge_array(left,right):
    result=[]
    i=0
    j=0
    n=len(left)
    m=len(right)
    while i<n and j<m:
        if left[i]<= right[j]:
            result.append(left[i])
            i+=1
        else :
            result.append(right[j])
            j+=1
        if i<n:
            while i<n:
                result.append[left[i]]
                i+=1
        if j<m:
            while j<m:
                result.append(right[j])
                j+=1
def sorted(arr):
    if len(arr)<=1:
        return arr
    mid=len(arr)//2
    left_arr=result[:mid]
    right_arr=result[mid:]
    left=sorted[left_arr]
    right=sorted[right_arr]
    return merge_array(left,right)
    

File: dsa in c/merge_sort.c
#include <stdio.h>
#include <stdlib.h>

int *merge(int left[], int right[], int n, int m)
{
    int *result = NULL;
    int size = 0;
    int i = 0, j = 0;

    // merge two arrays
    while (i < n && j < m)
    {
        size++;
        result = realloc(result, size * sizeof(int));

        if (left[i] <= right[j])
        {
            result[size - 1] = left[i++];
        }
        else
        {
            result[size - 1] = right[j++];
        }
    }

    // leftover from left side
    while (i < n)
    {
        size++;
        result = realloc(result, size * sizeof(int));
        result[size - 1] = left[i++];
    }

    // leftover from right side
    while (j < m)
    {
        size++;
        result = realloc(result, size * sizeof(int));
        result[size - 1] = right[j++];
    }

    return result;
}

int *merge_sort(int arr[], int n)
{
    if (n <= 1)
    {
        return arr;
    }

    int mid = n / 2;

    int *left = merge_sort(arr, mid);
    int *right = merge_sort(arr + mid, n - mid);

    int *result = merge(left, right, mid, n - mid);

    free(left);
    free(right);

    return result;
}

int main()
{
    int arr[] = {5, 2, 9, 1, 3, 8};
    int n = sizeof(arr) / sizeof(arr[0]);

    int *sorted = merge_sort(arr, n);

    for (int i = 0; i < n; i++)
        printf("%d ", sorted[i]);

    free(sorted);
    return 0;
}
